storage/storage_src.go

```diff
@@ -91,6 +98,18 @@ func (s *storageImageSource) Close() error {
 func (s *storageImageSource) GetBlob(ctx context.Context, info types.BlobInfo, cache types.BlobInfoCache) (rc io.ReadCloser, n int64, err error) {
 	// We need a valid digest value.
 	digest := info.Digest
+
+	digestIsTOC := false
+
+	// If the digest was overriden by LayerInfosForCopy, then we need to use the TOC digest
+	// to retrieve it from the storage.
+	s.getBlobMutex.Lock()
+	if v, ok := s.fromDigestToTOC[digest]; ok {
+		digest = v
+		digestIsTOC = true
```

Comment:
I’m not too happy with a variable that means “that other value with a clear name actually means something else entirely if this is set”.

But see elsewhere about storing a -> layer.ID mapping instead; that would make this distinction unnecessary.

At the very least, this can be moved below the `GzippedEmptyLayerDigest`, and then the this `if` and the `if digestIsTOC` below can be folded into a single condition.

Reply:
fixed